stream_alert/rule_processor/threat_intel.py

```diff
@@ -13,90 +13,389 @@
+                both normalized CEF type and IOC type.
+            (str): Second return is normalized type.
+            (str): Last return is IOC type.
         """
-        if cls.__intelligence:
-            return
-        if (config.get('threat_intel')
-                and config['threat_intel'].get('enabled')
-                and config['threat_intel'].get('mapping')):
-            cls.__intelligence = cls.read_compressed_files(intel_dir)
-            cls.__config = config['threat_intel'].get('mapping')
+        normalized_type = None
+        ioc_type = None
+
+        splitted_str = mapping_str.split(':')
+        if len(splitted_str) == 1:
+            normalized_type = splitted_str[0]
+        elif len(splitted_str) == 2:
+            normalized_type = splitted_str[0]
+            ioc_type = splitted_str[1].split('_')[-1]
+        else:
+            LOGGER.info('Key %s in conf/types.json is incorrect', mapping_str)
+            return False, None, None
+
+        if normalized_type and ioc_type:
+            return True, normalized_type, ioc_type
+
+        return False, normalized_type, None
+
+    def _process_ioc(self, ioc_collections):
+        """Check if any info is malicious by querying DynamoDB IOC table
+
+        Args:
+            ioc_collections (list): A list of StreamIoc instances.
+        """
+        # Segment data before calling DynamoDB table with batch_get_item.
+        for subset in StreamThreatIntel._segment(ioc_collections):
+            query_values = []
+            for ioc in subset:
+                if ioc.value not in query_values:
+                    query_values.append(ioc.value)
+            query_result = self._query(query_values)
+            for value in ioc_collections:
+                for ioc in query_result:
+                    if value.value.lower() == ioc[PRIMARY_KEY]:
+                        value.itype = ioc[SUB_TYPE_KEY]
+                        value.is_ioc = True
+                        continue
+
+    @staticmethod
+    def _segment(ioc_collections):
+        """Static method to segment ioc_collections in to smaller set.
+        Batch query to dynamodb supports up to 100 items.
+
+        Args:
+            ioc_collections (list): A list of StreamIoc instances
+
+        Returns:
+            (list): List of subset of StreamIoc instances
+        """
+        result = []
+        end = len(ioc_collections)
+        for index in range(0, end, MAX_QUERY_CNT):
+            result.append(ioc_collections[index:min(index+MAX_QUERY_CNT, end)])
+        return result
+
+    exceptions_to_backoff = (ClientError)
+    @backoff.on_exception(backoff.expo,
+                          exceptions_to_backoff,
+                          max_tries=3,
+                          giveup=exceptions_to_giveup,
+                          on_backoff=backoff_handler,
+                          on_success=success_handler,
+                          on_giveup=giveup_handler)
+    def _query(self, values):
```

Comment:
I would simplify this method to only have the batch_get_item request, and handle unprocessed keys by feeding them back into this function as input. The reason is that if for whatever reason either batch_get_item request fails, the entire request will retry with all values, which could be time consuming

Reply:
@jacknagz  Ideally, we should have a while loop to process unprocessed keys. But I am worry about the lambda timeout in Rule Processor. That's why I only re-query against `UnprocessedKeys` once now. We definitely can improve this in the near future. 
What do you think?

Reply:
Well, you technically wouldn't query it once, because backoff is applying to both requests.   My original comment still applies in this situation

Reply:
`backoff` will only be applied when there is `ClientError` exception. When the response includes `UnprocessedKeys`, it doesn't mean it returns error code. I would love to talk more offline.